iree/turbine/kernel/wave/scheduling/aplp/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PyTuple}; // PyAny removed as it's less relevant now
+use rayon::prelude::*;
+use std::sync::Arc;
+
+mod prune; // Contains the prune_envelope function
+
+// Define PyRawEdge as a tuple struct to directly accept Python tuples.
+// The order of fields in the Python tuple must match the order here:
+// (from_node, to_node, delay, iter_diff)
+#[derive(FromPyObject, Debug, Clone)]
+struct PyRawEdge(u32, u32, f32, f32);
+
+// Internal APLP computation function
+fn compute_aplp_internal(
+    node_count: usize,
+    raw_edges: Vec<PyRawEdge>, // Takes Vec of PyRawEdge (tuple structs)
+) -> Vec<Vec<Vec<(f32, f32)>>> {
+    let mut d_current_vec_vec: Vec<Vec<Vec<(f32, f32)>>> =
+        vec![vec![Vec::new(); node_count]; node_count];
+
+    for i in 0..node_count {
+        d_current_vec_vec[i][i] = prune::prune_envelope(vec![(0.0f32, 0.0f32)]);
+    }
+
+    for edge in raw_edges {
+        let u_idx = edge.0 as usize;
+        let v_idx = edge.1 as usize;
+        let delay = edge.2;
+        let iter_diff = edge.3;
+
+        if u_idx < node_count && v_idx < node_count {
+            let new_path_tuple = (delay, iter_diff);
+            let mut current_paths_at_uv = d_current_vec_vec[u_idx][v_idx].clone();
+            current_paths_at_uv.push(new_path_tuple);
+            d_current_vec_vec[u_idx][v_idx] = prune::prune_envelope(current_paths_at_uv);
+        }
+    }
+
+    let mut d_current_arc = Arc::new(d_current_vec_vec);
+
+    for k_idx in 0..node_count {
+        let d_next_rows: Vec<Vec<Vec<(f32, f32)>>> = (0..node_count)
+            .into_par_iter()
+            .map_with(Arc::clone(&d_current_arc), |d_curr_local, i_idx| {
+                let mut row_j_updates: Vec<Vec<(f32, f32)>> = vec![Vec::new(); node_count];
+                let paths_from_i_to_k = &d_curr_local[i_idx][k_idx];
+
+                for j_idx in 0..node_count {
+                    let paths_from_k_to_j = &d_curr_local[k_idx][j_idx];
+                    let mut candidate_paths_for_ij = d_curr_local[i_idx][j_idx].clone();
+
+                    if !paths_from_i_to_k.is_empty() && !paths_from_k_to_j.is_empty() {
+                        for &(d_ik, id_ik) in paths_from_i_to_k.iter() {
+                            if d_ik == -std::f32::INFINITY {
+                                continue;
+                            }
+                            for &(d_kj, id_kj) in paths_from_k_to_j.iter() {
+                                if d_kj == -std::f32::INFINITY {
+                                    continue;
+                                }
+                                candidate_paths_for_ij.push((d_ik + d_kj, id_ik + id_kj));
+                            }
+                        }
+                    }
+                    row_j_updates[j_idx] = prune::prune_envelope(candidate_paths_for_ij);
+                }
+                row_j_updates
```

Comment:
is this dead code? (not sure since I am new to rust)

Reply:
or is this a return from this lambda function inside `map_with` region?

Reply:
Good question. So this is a closure (https://doc.rust-lang.org/rust-by-example/fn/closures.html) which you can think of as a lambda in C++. This is the return value of the closure which is the last line in the closure without a semicolon (https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch13-01-closures.html#:~:text=The%20value%20returned%20from%20the,just%20like%20in%20function%20bodies.)